keps/262-ConfigurableFailurePolicy/README.md

```diff
@@ -0,0 +1,392 @@
+    - [Unit Tests](#unit-tests)
+    - [Integration tests](#integration-tests)
+  - [Graduation Criteria](#graduation-criteria)
+- [Implementation History](#implementation-history)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+
+## Motivation
+
+JobSet's current FailurePolicy API only allows specifying the max number of JobSet restarts to attempt before marking it as failed.
+
+Users have requested that we should also allow them to configure the failure conditions under which the JobSet will be restarted,
+or when the JobSet should be failed immediately without going through the allowed number of restarts. 
+
+This will allow users to make more efficient use of their computing resources, by avoiding `maxRestarts` number of unnecessary 
+JobSet restarts in the case of a non-retriable failure (e.g., an application code bug), while still restarting in the event of
+retriable failures (e.g., a host maintenace event).
+
+This is especially important for large scale distributed training workloads on expensive hardware accelerators. Recreating the 
+entire JobSet may take ~1 minute or so (per latest scale tests on several node pools of ~3k nodes each), and if `maxRestarts`
+is set to a very high number to allow for disruptions (which occur often at a large scale like this), the workload could
+waste roughly `maxRestarts` number of minutes repeatedly recreating despite the fact that it is doomed to fail due to a non-retriable
+error like an application code bug. This is very expensive and wasteful on expensive, large scale clusters.
+
+We need to define a more configurable Failure Policy API which will allow users to define their own restart and failure semantics, 
+based on the type of child job failure occurring.
+
+### Goals
+
+- Enhance the Failure Policy API with options allowing the user to control the restart and failure semantics of the JobSet.
+
+### Non-Goals
+
+- Handling every possible type of restart and failure semantics. We just want to provide enough flexibility to handle common use
+cases.
+
+## Proposal
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a user, in order to use my computing resources as efficiently as possible, I want to 
+configure my JobSet to restart in the event of a child job failure due to a retriable error
+like host maintance, but to fail the JobSet immediately without any unnecessary restart 
+attempts in the event of an non-retriable application code error. 
+
+**Example Failure Policy Configuration for this use case**:
+
+```yaml
+apiVersion: jobset.x-k8s.io/v1alpha2
+kind: JobSet
+metadata:
+  name: configurable-failure-policy-1
+spec:
+  # Failure Policy configured to fail the JobSet immediately if a job failure reason for any child job was NOT due to a SIGTERM (e.g. host maintenance).
+  # Otherwise, restart up to 10 times.
+  failurePolicy:
+    rules:
+    - action: FailJobSet
+      onJobFailureReasons: 
+      - PodFailurePolicy
```

Comment:
I think it would be great if we can send an upstream feature request to batchv1.Job to optionally allow setting the reason for the podFailurePolicy as part of the action in that api, or if we don't want to change the api, in the case of onExitCodes to have the code appended to the reason, like "PodFailurePolicy-ExitCode143".

This will give users of jobSet more flexibility since they may want the failure policy to be configured differently depending on the exact reason of pod failure (even if in all cases that caused the job to fail).

Reply:
@liggitt @alculquicondor what do you think? Which option is more realistic to pursue?

Reply:
For the HPC use case it would be hugely useful to be able to act on cases down to the exit code.

Reply:
Yes indeed! From JobSet perspective, I think we have the right API. The change needs to happen in the Job controller since it is the one dealing with the pods.

Reply:
Created https://github.com/kubernetes/kubernetes/issues/122972 to track the issue

Reply:
I think separate KEP is preferable as we would like to go GA in 1.31 with pod failure policy, which is already incubating for a while, and IIUC the new API field should be protected by Alpha level feature gate first to work properly when the API servers are in skewed versions.

In the meanwhile you could also consider matching the `message` field in the condition added by pod failure policy. It contains details information about the exit code, or pod condition, see: https://github.com/kubernetes/kubernetes/blob/7340ce932bf3071c73ca371e4dfdd5ba00365668/pkg/controller/job/pod_failure_policy_test.go#L86